### PR TITLE
Add section Understanding Quantum Teleportation to the chapter

### LINKED ISSUE #618: Use teleportation to deepen understanding of quantum concepts
**Is your feature request related to a problem? Please describe.**
The teleportation chapter currently describes what quantum teleportation is, and provides an example. The protocol is not explained, and this does not teach the reader anything new about quantum mechanics.

**Describe the solution you'd like**
I would like to see the teleportation chapter include an explanation of how the protocol works (as suggested in #382), but also to go further and use the chapter to teach / consolidate quantum concepts. In particular, I would like to show that (assuming `|psi> = a|0> + b|1>`), measuring the two qubits projects any superposition onto a state containing both amplitudes `a` and `b`. E.g:

`1/sqrt(2)( a|1>|10> + b|0>|10> )`

From which we can regain the teleported state through a combination of `X` and `Z` gates. This effect is seen in Simon's algorithm which currently follows shortly after, so this could be a good stepping stone.

If this goes well, we can perhaps even move the chapter to chapter 2 as an early example. (Including examples earlier has been suggested by multiple sources).

**Describe alternatives you've considered**
Alternatively, we can go the opposite direction and turn this into an exercise, where we guide the reader through working this out for themselves. Exactly how this might fit in with the labs and current exercises is not clear.

### PR DESCRIPTION
**Pull request by rochisha0**
# Changes made

I intend to add a section which touches upon the quantum concepts of teleportation.
It fixes #618 

# Justification

It is important to explain the mathematics behind the protocol. The same has been covered for Superdense Coding but not Quantum Teleportation,

### CODE REVIEW

File: content/ch-algorithms/teleportation.ipynb [hunk 1]
```diff
@@ -22411,6 +22411,66 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "As you have worked with the Quantum Teleportation's implementation, it is time to understand the mathematics behind the protocol.\n",
+    "\n",
+    "#### Step 1\n",
+    "\n",
+    "Quantum Teleportation begins with the fact that Alice needs to transmit $|\\psi\\rangle = a|0\\rangle + b|1\\rangle$ (a random qubit) to Bob. She doesn't know the state of the qubit. For this, Alice and Bob take the help of a third party (Telamon). Telamon prepares a pair entangled qubits for Alice and Bob. The entangled qubit could be written in Dirac Notation as,\n",
```
Comment (aditya-giri) [inline content/ch-algorithms/teleportation.ipynb:22429]: Telamon prepares a pair entangled qubits --> Telamon prepares a pair _of_ entangled qubits

File: content/ch-algorithms/teleportation.ipynb [hunk 2]
```diff
@@ -22411,6 +22411,66 @@
...
+    "$$ |q \\rangle = \\frac{1}{\\sqrt{2}} (|00\\rangle + |11\\rangle) $$\n",
+    "\n",
+    "Alice and Bob each possess one qubit of the entangled pair (denoted as A and B respectively),\n",
+    "$$|\\psi\\rangle = \\frac{1}{\\sqrt{2}} (|0\\rangle_A |0\\rangle_B + |1\\rangle_A |1\\rangle_B) $$\n",
```
Comment (aditya-giri) [inline content/ch-algorithms/teleportation.ipynb:22433]: I think this state should be _q_, not _psi_?

**Comment by aditya-giri** [comment]
I believe it would make more sense to put this section _before_ section 3: simulation? Right now it's sitting in between the simulation section and the section on running on a real backend.

**Comment by rochisha0** [comment]
Hey, @aditya-giri ! Thanks for your comments, it is a still a work in progress! With the part of it being in the 4th section is @frankharkins idea, actually. Checkout the issue #618 for more details.

**Comment by frankharkins** [comment]
I don't have a strong preference on where the content goes, I imagined a new section 4, but see arguments for all positions. To clarify, @rochisha0 you are free to decide where you think the content fits best.

**Comment by rochisha0** [comment]
Hi @frankharkins ! I think before the third section it will be a bit problematic because it will introduce the heavy mathematics straight.

File: content/ch-algorithms/teleportation.ipynb [hunk 3]
```diff
@@ -22415,14 +22415,82 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Step 2\n",
+    "\n",
+    "Now according to the protocol Alice applies CNOT gate on her two qubits followed by Hadamard gate on the first qubit. This results in the state,\n",
+    "\n",
+    "$$ \\begin{align} (CNOT \\otimes I) (H \\otimes I \\otimes I) (|q\\rangle \\otimes |\\psi\\rangle) \\\\\n",
```
Comment (frankharkins) [inline content/ch-algorithms/teleportation.ipynb:22450]: I'm not sure about this; shouldn't the (CNOT x I)  be on the right of the (H x I x I ) since it is applied first?

File: content/ch-algorithms/teleportation.ipynb [hunk 4]
```diff
@@ -22415,14 +22415,82 @@
...
+    "&=(CNOT \\otimes I) (H \\otimes I \\otimes I) \\frac{1}{\\sqrt{2}} (a|000\\rangle + a|011\\rangle + b|100\\rangle + b|111\\rangle) \\\\\n",
+    "&= (H \\otimes I) \\frac{1}{\\sqrt{2}} (a|000\\rangle + a|011\\rangle + b|110\\rangle + b|101\\rangle) \\\\\n",
```
Comment (frankharkins) [inline content/ch-algorithms/teleportation.ipynb:22452]: Also should this be (H x I x I) instead of (H x I)?

File: content/ch-algorithms/teleportation.ipynb [hunk 5]
```diff
@@ -22415,14 +22415,82 @@
...
+    "&= \\frac{1}{2}  (a(|000\\rangle + |011\\rangle + |100\\rangle + |111\\rangle) + b(|010\\rangle + |001\\rangle - |110\\rangle - |101\\rangle)) \\\\\n",
+    "&= \\frac{1}{2} ( |00\\rangle (a|0\\rangle + b|1\\rangle) + |01\\rangle (a|1\\rangle + b|0\\rangle) +|10\\rangle (a|0\\rangle - b|1\\rangle) +|11\\rangle (a|1\\rangle - b|0\\rangle) )\n",
```
Comment (frankharkins) [inline content/ch-algorithms/teleportation.ipynb:22454]: This is subjective, but I would personally prefer to break this over 4 lines. I.e.
 ```
\\frac{1}{2} (  |00\\rangle (a|0\\rangle + b|1\\rangle) 
              + |01\\rangle (a|1\\rangle + b|0\\rangle) 
              +|10\\rangle (a|0\\rangle - b|1\\rangle) 
              +|11\\rangle (a|1\\rangle - b|0\\rangle))
```

File: content/ch-algorithms/teleportation.ipynb [hunk 6]
```diff
@@ -22415,14 +22415,82 @@
...
+    "\\end{align}$$"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Step 3\n",
+    "\n",
+    "Alice measures the first two qubit (which she owns) and sends them as two classical bits to Bob. The result she obtains is always one of the four standard basis states $|00\\rangle, |01\\rangle, |10\\rangle,$ and $|11\\rangle$ with equal probability.  \n",
```
Comment (frankharkins) [inline content/ch-algorithms/teleportation.ipynb:22464]: Related to[ this suggestion](https://github.com/qiskit-community/qiskit-textbook/pull/651/files#r469926240), it would be cool to see a table-like equation here showing:
```
|00>  --->  a|0> + b|1>
|01>  --->  b|1> + a|0> 
etc.
```
[Similar to this equation here](https://qiskit.org/textbook/ch-algorithms/teleportation.html#Step-4).

**Comment by frankharkins** [comment]
@rochisha0 thanks for writing this, I have just had time to review. I find step 1 very clear and concise, and I have left some comments on steps 2 & 3. I have not commented on step 4 as I see you are still working on it. Please let me know what you think.

File: content/ch-algorithms/teleportation.ipynb [hunk 7]
```diff
@@ -24415,7 +24485,7 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.8.3"
+   "version": "3.6.9"
```
Comment (aditya-giri) [inline content/ch-algorithms/teleportation.ipynb:24488]: you should probably revert the change on this line right before you check it in?

File: content/ch-algorithms/teleportation.ipynb (hunk 3 above)
Comment (aditya-giri) [inline content/ch-algorithms/teleportation.ipynb:22450]: I'm not able to see these lines on the rendered file within github (https://github.com/rochisha0/qiskit-textbook/blob/update_qt/content/ch-algorithms/teleportation.ipynb). Is something missing? The previous lines seem to be displayed okay...

File: content/ch-algorithms/teleportation.ipynb (hunk 4 above)
Comment (aditya-giri) [reply content/ch-algorithms/teleportation.ipynb:22452]: Yes, it would be better to write out all the gates in terms of their actions on all 3 qubits just to avoid any confusion later...

File: content/ch-algorithms/teleportation.ipynb (hunk 7 above)
Comment (rochisha0) [reply content/ch-algorithms/teleportation.ipynb:24488]: I don't think this makes much difference. Thanks

File: content/ch-algorithms/teleportation.ipynb (hunk 6 above)
Comment (rochisha0) [reply content/ch-algorithms/teleportation.ipynb:22464]: Yes, I'm working on this only currently. :)

File: content/ch-algorithms/teleportation.ipynb (hunk 5 above)
Comment (rochisha0) [reply content/ch-algorithms/teleportation.ipynb]: Hey! It doesn't render perfectly on LaTeX.

**Comment by rochisha0** [comment]
Hey @frankharkins  and @aditya-giri !  I have finalized step 4 from my side, please review :)

File: content/ch-algorithms/teleportation.ipynb (hunk 3 above)
Comment (rochisha0) [reply content/ch-algorithms/teleportation.ipynb]: I'm not sure, everything looks fine in the local clone.

**Comment by frankharkins** [comment]
@rochisha0, step 4 looks good. I had a look at the latex as I remember having similar problems in the past, and think I've found an acceptable workaround. [I've made a PR here for you to review](https://github.com/rochisha0/qiskit-textbook/pull/1).

**Comment by rochisha0** [comment]
It is an interesting way to implement the equation with latex. Thanks @frankharkins !

**Comment by frankharkins** [comment]
I'm pleased with this. @rochisha0 are you happy for me to merge?

**Comment by rochisha0** [comment]
Yes, @frankharkins ! I'm excited for this to get merged! Thank you!
